src/MongoDB.Driver.Core/Core/Misc/SemanticVersion.cs

```diff
@@ -106,11 +130,16 @@ public string PreRelease
         /// <inheritdoc/>
         public int CompareTo(SemanticVersion other)
         {
-            if (other == null)
+            if (ReferenceEquals(other, null))
             {
                 return 1;
             }
 
+            if (IsInternalServerBuild() || other.IsInternalServerBuild())
+            {
+                return this.AsServerVersion().CompareTo(other.AsServerVersion());
+            }
+
```

Comment:
I think your idea from an earlier commit to place this check AFTER comparing major, minor and patch is a good idea.

Go ahead and move it if you agree.